File: Editorial/42_Trapping_Rain_Water/Go/3_Using_stacks.go
func trap(height []int) int {
    ans := 0
    current := 0
    var st []int = make([]int, 0)
    for current < len(height) {
        for len(st) != 0 && height[current] > height[st[len(st)-1]] {
            top := st[len(st)-1]
            st = st[:len(st)-1]
            if len(st) == 0 {
                break
            }
            distance := current - st[len(st)-1] - 1
            bounded_height := min(
                height[current],
                height[st[len(st)-1]],
            ) - height[top]
            ans += distance * bounded_height
        }
        st = append(st, current)
        current++
    }
    return ans
}

func min(a int, b int) int {
    if a < b {
        return a
    } else {
        return b
    }
}

File: Editorial/42_Trapping_Rain_Water/C/3_Using_stacks.c
int trap(int* height, int heightSize) {
    int ans = 0, current = 0;
    int* st = malloc(heightSize * sizeof(int));
    int stSize = 0;
    while (current < heightSize) {
        while (stSize != 0 && height[current] > height[st[stSize - 1]]) {
            int top = st[stSize - 1];
            stSize--;
            if (stSize == 0) break;
            int distance = current - st[stSize - 1] - 1;
            int bounded_height =
                fmin(height[current], height[st[stSize - 1]]) - height[top];
            ans += distance * bounded_height;
        }
        st[stSize++] = current++;
    }
    free(st);
    return ans;
}
